src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedEntryPointSymbol.cs

```diff
@@ -329,6 +335,104 @@ private static BoundCall CreateParameterlessCall(CSharpSyntaxNode syntax, BoundE
             { WasCompilerGenerated = true };
         }
 
+        /// <summary> A synthesized entrypoint that forwards all calls to an async Main Method </summary>
+        internal sealed class AsyncForwardEntryPoint : SynthesizedEntryPointSymbol
+        {
+            /// <summary> The user-defined asynchronous main method. </summary>
+            private readonly CSharpSyntaxNode _userMainReturnTypeSyntax;
+
+            private readonly BoundExpression _getAwaiterGetResultCall;
+
+            private readonly ImmutableArray<ParameterSymbol> _parameters;
+
+            internal AsyncForwardEntryPoint(CSharpCompilation compilation, DiagnosticBag diagnosticBag, NamedTypeSymbol containingType, MethodSymbol userMain) :
+                base(containingType)
+            {
+                // There should be no way for a userMain to be passed in unless it already passed the 
+                // parameter checks for determining entrypoint validity.
+                Debug.Assert(userMain.ParameterCount == 0 || userMain.ParameterCount == 1);
+
+                _userMainReturnTypeSyntax = userMain.ExtractReturnTypeSyntax();
+                // PROTOTYPE(async-main): we might need to adjust the containing member of the binder to be the Main method.
+                var binder = compilation.GetBinder(_userMainReturnTypeSyntax);
+                _parameters = SynthesizedParameterSymbol.DeriveParameters(userMain, this);
+
+                var arguments = Parameters.SelectAsArray(p => (BoundExpression)new BoundParameter(_userMainReturnTypeSyntax, p, p.Type));
```

Comment:
Minor point: Use `SelectAsArray` overload with additional argument to avoid a closure class:
```
Parameters.SelectAsArray((p, s) => (BoundExpression)new BoundParameter(s, p, p.Type), _userMainReturnTypeSyntax)
```